File: core/src/main/java/forpleuvoir/dhwu1a/core/Dhwu1a.java
package forpleuvoir.dhwu1a.core;

import forpleuvoir.dhwu1a.config.Dhwu1aConfig;
import forpleuvoir.dhwu1a.config.LogConfig;
import forpleuvoir.dhwu1a.core.event.base.EventBus;
import forpleuvoir.dhwu1a.core.user.bot.Bot;

import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.function.Consumer;

/**
 * 程序入口
 *
 * @author forpleuvoir
 * <p>#project_name dhwu1a
 * <p>#package forpleuvoir.dhwu1a.core
 * <p>#class_name Dhwu1a
 * <p>#create_time 2021/6/28 20:28
 */
public class Dhwu1a {
    private static Dhwu1a INSTANCE;

    public final LogConfig logConfig;
    public Bot bot;
    private boolean running;
    private final Dhwu1aConfig config;
    private EventBus eventBus;

    private final List<Consumer<Dhwu1a>> onStart = new CopyOnWriteArrayList<>();
    private final List<Consumer<Dhwu1a>> onClose = new CopyOnWriteArrayList<>();
    private final List<Consumer<Dhwu1a>> onRestart = new CopyOnWriteArrayList<>();


    public static Dhwu1a getInstance() {
        return INSTANCE;
    }

    public static void initialize(Dhwu1aConfig config) {
        if (INSTANCE == null) {
            INSTANCE = new Dhwu1a(config);
        }
    }

    private Dhwu1a(Dhwu1aConfig config) {
        Thread.currentThread().setName("dhwu1a");
        this.logConfig = config.getLogConfig();
        this.config = config;
    }

    public void start() {
        this.eventBus = EventBus.getInstance();
        Bot.initialize(config);
        this.bot = Bot.getInstance();
        running = true;
        onStart();
    }

    private void onStart() {
        onStart.forEach(onStartConsumer -> onStartConsumer.accept(this));
    }

    public void reStart() {
        start();
        onRestart();
    }

    private void onRestart() {
        onRestart.forEach(onRestartConsumer -> onRestartConsumer.accept(this));
    }

    public void close() {
        bot.close();
        bot = null;
        running = false;
        onClose();
    }

    private void onClose() {
        onClose.forEach(onCloseConsumer -> onCloseConsumer.accept(this));
    }

    public boolean isRunning() {
        return running;
    }

    public Bot getBot() {
        return bot;
    }


    public void addOnStartListener(Consumer<Dhwu1a> consumer) {
        onStart.add(consumer);
    }

    public void addOnRestartListener(Consumer<Dhwu1a> consumer) {
        onRestart.add(consumer);
    }

    public void addOnCloseListener(Consumer<Dhwu1a> consumer) {
        onClose.add(consumer);
    }

    public EventBus getEventBus() {
        return eventBus;
    }
}

File: core/src/main/kotlin/forpleuvoir/dhwu1a/core/Dhwu1a.kt
package forpleuvoir.dhwu1a.core

import forpleuvoir.dhwu1a.core.config.Dhwu1aConfig
import forpleuvoir.dhwu1a.core.config.LogConfig
import forpleuvoir.dhwu1a.core.event.base.EventBus
import forpleuvoir.dhwu1a.core.user.bot.Bot
import forpleuvoir.dhwu1a.core.util.Dhwu1aLog
import java.util.concurrent.ConcurrentLinkedDeque

/**
 * 程序入口
 *
 * @author forpleuvoir
 *
 * #project_name dhwu1a
 *
 * #package forpleuvoir.dhwu1a.core
 *
 * #class_name Dhwu1a
 *
 * #create_time 2021/6/28 20:28
 */
class Dhwu1a private constructor() {
    companion object {
        @JvmStatic
        val instance: Dhwu1a by lazy { Dhwu1a() }
    }

    private lateinit var config: Dhwu1aConfig
    lateinit var bot: Bot
        private set
    var isRunning = false
        private set
    val eventBus: EventBus = EventBus.instance
    private val onStart: MutableCollection<(Dhwu1a) -> Unit> = ConcurrentLinkedDeque()
    private val onClose: MutableCollection<(Dhwu1a) -> Unit> = ConcurrentLinkedDeque()
    private val onRestart: MutableCollection<(Dhwu1a) -> Unit> = ConcurrentLinkedDeque()

    private val log: Dhwu1aLog = Dhwu1aLog(this::class.java)

    fun initialize(config: Dhwu1aConfig) {
        Thread.currentThread().name = "dhwu1a"
        LogConfig.instance.copyOf(config.logConfig)
        this.config = config
        this.bot = Bot.instance
    }

    fun start() {
        printSystemInfo()
        bot.initialize(config)
        isRunning = true
        onStart()
    }

    private fun printSystemInfo() {
        val javaVersion = System.getProperty("java.version")
        val osName = System.getProperty("os.name")
        val osVersion = System.getProperty("os.version")
        val userName = System.getProperty("user.name")
        log.info("JVM版本:{}", javaVersion)
        log.info("用户名:{}", userName)
        log.info("运行系统:{}", osName)
        log.info("系统版本:{}", osVersion)
    }

    private fun onStart() {
        onStart.forEach { it.invoke(this) }
    }

    fun reStart() {
        start()
        onRestart()
    }

    private fun onRestart() {
        onRestart.forEach { it.invoke(this) }
    }

    fun close() {
        bot.close()
        isRunning = false
        onClose()
    }

    private fun onClose() {
        onClose.forEach { it.invoke(this) }
    }

    fun addOnStartListener(listener: (Dhwu1a) -> Unit) {
        onStart.add(listener)
    }

    fun addOnRestartListener(listener: (Dhwu1a) -> Unit) {
        onRestart.add(listener)
    }

    fun addOnCloseListener(listener: (Dhwu1a) -> Unit) {
        onClose.add(listener)
    }


}